### PR TITLE
Invalidate transaction as early as possible

### PR DESCRIPTION
**Pull request by nvasilevski**
### Motivation / Background

This PR changes the way of invalidating a transaction for two reasons
- Avoid `Mysql2::Error: SAVEPOINT active_record_%n% does not exist` error being raised when application wrongly handles `ActiveRecord::Deadlocked` exception by manually trying to rollback transaction that has already been rolled back by the database itself
- Invalidating transaction as early as possible is a better design as it moves the transaction in a correct state as soon as possible so other parts of the framework can rely on that. I'll provide an example below

### Detail

- The main motivation behind this chance is to reduce the number of `Mysql2::Error: SAVEPOINT active_record_%n% does not exist`. I have to admit that this error is happening mainly because of the wrong design of an application or specifically transaction handling in an application. The test shows a simplified version of the design, generally it looks like: 1. start a transaction 2. define general `rescue` in the transaction block which basically rescues anything that inherits from `StandardError` 3. handle the exception by logging/submitting stats/ etc and raise an `ActiveRecord::Rollback` or basically any exception that causes framework to issue a `ROLLBACK` statement leading to a `Mysql2::Error` because there is nothing to rollback on the db side but framework can't know this just yet and the invalidation is only about to happen here 
https://github.com/rails/rails/blob/c8315a96f962b0702062197c5a37d6d0a4bc9bc8/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L459-L462
which is too late and happens after we do our own `rescue`-`raise` flow causing the `ROLLBACK` to be issued without necessity
It the transaction was invalidated earlier, as proposes this PR, the rollback would have been ignored:
https://github.com/rails/rails/blob/c8315a96f962b0702062197c5a37d6d0a4bc9bc8/activerecord/lib/active_record/connection_adapters/abstract/transaction.rb#L276-L278
This may sound as a purely application problem and the simplicity of the test may hide the fact that in reality applications are much more complex and those steps 1. open transaction 2. rescue 3. raise/rollback are widely spread and it's not always easy to refactor especially with no proper domain context or even realize that as a developer that you are wrongly handling transactions.
I consider the knowledge about how deadlock behaves in mysql and how properly handle exceptions within a transaction to be closer to an advanced level which in my opinion justifies the proposal of making framework a bit smarter and a bit forgiving so it can prevent wrong thing from happening. Especially because we already, by design, do not issue `ROLLBACK` statements for a transaction with an invalidated state. This PR just helps to handle it in the same way in more cases.
Even if we can't consider my first point as a sufficient argument, I have the second one:

- Second motivation for this PR is that I strongly believe that invalidating a transaction as early as possible is a better design and "just makes sense". As I mentioned above, framework already makes certain decisions based on the transaction state so keeping the state as correct as possible is the right thing. 
For example after properly invalidating transaction state, we should be able to get rid of the following check - 
https://github.com/rails/rails/blob/999c58d41fe4726ef2b90910cea934f0593aad49/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L1021
Specifically I strongly believe that `savepoint_errors_invalidate_transactions?` check should only be used for deciding whether a transaction should be invalidated or not. If the check is used in a different context it only means that we are checking for something by implication and we lack a more specific check. So I'll be also proposing refactoring this check to replace it with something like (pseudo code) `do_not_retry if current_transaction && current_transaction.state.invalidated?` which will basically make this comment redundant as the code will be self-explanatory
https://github.com/rails/rails/blob/999c58d41fe4726ef2b90910cea934f0593aad49/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L1019-L1020

### What's next

- There is also a case when `COMMIT` statement can be ignored as well for an invalidated transaction. I need to double check it and I'll prepare a PR if it's still an issue
- As I mentioned above, I'd like to simplify this check https://github.com/rails/rails/blob/999c58d41fe4726ef2b90910cea934f0593aad49/activerecord/lib/active_record/connection_adapters/abstract_adapter.rb#L1021 as after this PR it should be unnecessary to perform all three checks as we can simply check the transaction state

### Checklist

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [x] CI is passing.

### CODE REVIEW

**Comment by matthewd** [comment]
Nice! I'll tell myself that prior to `with_raw_connection` there wasn't a similarly fitting place to put invalidation closer to the initial failure 😅 

> For example after properly invalidating transaction state, we should be able to get rid of the following check

Yeah, that sounds much better.

> making framework a bit smarter and a bit forgiving

TBH, I've been wondering if we should do even more here. Right now we correctly track whether there's really a transaction still open on the DB side, but we don't interfere if the app tries to run a query. I feel like I could easily make the argument that if your Ruby code is nested in a situation where it could/should reasonably believe that it's wrapped by a transaction, but the transaction has in fact disappeared, we should refuse to talk to the server until the code has unwound to a point where we have officially rolled back on our side. 🤔

File: activerecord/test/cases/adapters/mysql2/nested_deadlock_test.rb [hunk 1]
```diff
@@ -78,6 +78,57 @@ class Sample < ActiveRecord::Base
       assert_predicate connection, :active?
     end
 
+    test "rollback exception is swallowed after a rollback" do
```
Comment (eileencodes) [inline activerecord/test/cases/adapters/mysql2/nested_deadlock_test.rb:81]: Is this specific to mysql or should we also include a test in test/cases/adapters/postgresql/transaction_nested_test.rb for postgres?

**Review by eileencodes — COMMENTED** [review]
I think this change should have a CHANGELOG entry for Active Record.

File: activerecord/test/cases/adapters/mysql2/nested_deadlock_test.rb (hunk 1 above)
Comment (nvasilevski) [reply activerecord/test/cases/adapters/mysql2/nested_deadlock_test.rb:81]: At this point applies to `mysql2` adapter only

But I would put my answer as "it only applies to adapters that have `savepoint_errors_invalidate_transactions` configured as `true`", which, by implication, scopes the change to mysql only

https://github.com/rails/rails/blob/682353e3dea9bead97c5fe83b799f65723954412/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb#L78-L80

**Comment by nvasilevski** [comment]
Added a unit test as the one with real deadlock looks too integration-like

**Pull request merged**